File: android/src/main/java/com/cas/extensions/MediationManagerWrapper.kt
package com.cas.extensions

import com.cleversolutions.ads.AdLoadCallback
import com.cleversolutions.ads.AdType
import com.cleversolutions.ads.MediationManager
import java.util.*
import kotlin.collections.HashMap

typealias MediationManagerListener = (manager: MediationManager?) -> Unit

class MediationManagerWrapper: AdLoadCallback {
  private val managerChangeListeners = HashMap<String, MediationManagerListener>()
  private val adLoadListeners = HashMap<String, AdLoadCallback>()

  var manager: MediationManager? = null
    set(value) {
      field = value
      managerChangeListeners.values.forEach { it(value) }
      field?.onAdLoadEvent?.add(this)
    }

  fun addChangeListener(listener: MediationManagerListener): String {
    val key = UUID.randomUUID().toString()
    managerChangeListeners[key] = listener
    return key
  }

  fun removeChangeListener(key: String) {
    managerChangeListeners.remove(key)
  }

  fun addLoadListener(listener: AdLoadCallback): String {
    val key = UUID.randomUUID().toString()
    adLoadListeners[key] = listener
    return key
  }

  fun removeLoadListener(key: String) {
    adLoadListeners.remove(key)
  }

  override fun onAdFailedToLoad(type: AdType, error: String?) {
    adLoadListeners.values.forEach { it.onAdFailedToLoad(type, error) }
  }

  override fun onAdLoaded(type: AdType) {
    adLoadListeners.values.forEach { it.onAdLoaded(type) }
  }
}

File: ios/MediationManagerWrapper.swift
//
//  MediationManagerWrapper.swift
//  react-native-cas
//
//  Created by Lonely Astronaut on 20.09.23.
//

import Foundation
import CleverAdsSolutions

typealias MediationManagerListener = (_ manager: CASMediationManager) -> Void

class MediationManagerWrapper: CASLoadDelegate {
    private init() {}
    static let shared = MediationManagerWrapper()

    private var managerChangeListeners: Dictionary<String, MediationManagerListener> = [:]
    private var adLoadListeners: Dictionary<String, CASLoadDelegate> = [:]
    
    var manager: CASMediationManager? = nil {
        didSet {
            if manager != nil {
                managerChangeListeners.values.forEach { $0(manager!) }
                manager!.adLoadDelegate = self
            }
        }
    }
    
    func addChangeListener(_ listener: @escaping MediationManagerListener) -> String {
        let key = UUID().uuidString
        managerChangeListeners[key] = listener
        return key
    }
    
    func removeChangeListener(_ key: String) {
        managerChangeListeners.removeValue(forKey: key)
    }
    
    func addLoadListener(_ listener: CASLoadDelegate) -> String {
        let key = UUID().uuidString
        adLoadListeners[key] = listener
        return key
    }
    
    func removeLoadListener(_ key: String) {
        adLoadListeners.removeValue(forKey: key)
    }
    
    func onAdLoaded(_ adType: CleverAdsSolutions.CASType) {
        adLoadListeners.values.forEach{ $0.onAdLoaded(adType) }
    }
    
    func onAdFailedToLoad(_ adType: CleverAdsSolutions.CASType, withError error: String?) {
        adLoadListeners.values.forEach{ $0.onAdFailedToLoad(adType, withError: error) }
    }
}
